### DISCUSSION TITLE
Replace  `createWithDurations` with `createWithTimestamps` in Airstreams

### DISCUSSION THREAD

**Discussion by smol-ninja** [sablier-labs/evm-monorepo#997] [Ideas]
## Offers
1. **[One Time Airdrop](https://github.com/sablier-labs/company-discussions/discussions/55)**: Although more expensive than having dedicated one-time airdrop logic, it allows `claim` and `withdraw` in a single tx.
2. **[Ranged Airstreams](https://github.com/sablier-labs/v2-periphery/discussions/279)**: This implementation supports airstreams with uniform start and end times.
3. **Legacy airstreams**[^1]

## Context

Relying on `createWithDurations` in `MerkleLockup` has the following limitations:
1. Airstreams cannot start and end at the same time regardless of when they are claimed.
2. Airstreams cannot be used for one-time airdrop because [the end time must be greater than the start time](https://github.com/sablier-labs/v2-core/blob/daeac7c23942ac465fd5da5d8291ecd433326e72/src/libraries/Helpers.sol#L171-L173), preventing token withdrawal in the same transaction.

It can only be used to create a campaign where streams begin vesting only when claimed. But what if we can also achieve the same functionality using `createWithTimestamps`?

## Proposal

**Replace `createWithDurations` with `createWithTimestamps`**. This would require replacing [`streamDurations`](https://github.com/sablier-labs/v2-periphery/blob/68b5cfab1a2a0912bcf85745de2cd98406181c7f/src/SablierV2MerkleLL.sol#L32C44-L32C59) with a bespoke timestamp variable but with a slight modification:

```solidity
MerkleLockup.Timestamps public override timestamps;

MerkleLockup.Timestamps {
    uint40 startTime;
    uint40 cliffDuration;
    uint40 endDuration;
}
```

When a user claims, calculate `cliffTime` and `endTime` as follows to create **ranged airstreams**:
- cliffTime = startTime + cliffDuration
- endTime = startTime + endDuration

I will explain below why it looks different than [LockupLinear.Timestamp](https://github.com/sablier-labs/v2-core/blob/daeac7c23942ac465fd5da5d8291ecd433326e72/src/types/DataTypes.sol#L275-L279).

### One-time airdrops

This can be achieved by setting the end time in the past. But since Lockup [requires end time to be in the future](https://github.com/sablier-labs/v2-core/blob/daeac7c23942ac465fd5da5d8291ecd433326e72/src/libraries/Helpers.sol#L175-L179), we will have to change that and allow end time to be in the past as long as the following inequality holds: `startTime < cliffTime < endTime`.

> [!NOTE] 
 > Even though, one-time airdrop can be unlocked with this approach, I am not against having a bespoke contract for it (see [discussion sablier-labs/v2-periphery#370](https://github.com/sablier-labs/v2-periphery/discussions/370)). Its a similar argument as to why we chose to release Lockup Tranched even though the same curve can also be achieved by Lockup Dynamic - gas efficiency.

### Legacy Airstreams

To create the same airstreams that we allow today, we will allow `startTime` to be 0. When its 0, Lockup Linear's `startTime` will default to `block.timestamp`, making `cliffTime` and `endTime`:
- cliffTime = block.timestamp + cliffDuration
- endTime = block.timestamp + endDuration

This would create airstreams that begin vesting when the user claims and end at different dates, similar to the current behaviour.

cc @sablier-labs/engineers.

<img width="356" alt="Screenshot 2024-07-19 at 12 01 10" src="https://github.com/user-attachments/assets/cfa604d8-314d-4f84-b495-314ab0986030">

[^1]: Legacy airstreams are the current airstreams supported by Merkle Lockup. They have start time set as claim time and end time at a fixed duration.

**Comment by andreivladbrg** [comment]
Before giving a full response, I want to mention there is a security reason for using `createWithDuration` over `createWithTimestamps`. With `createWithTimestamps`, if `block.timestamp > endTime`, claiming would become frozen as we need to have an end time greater than the current time. For a campaign that has no expiration, this means those funds would become stuck.

**Reply by smol-ninja** [reply]
Yes. You are right and that's why I included the following suggestion in the OP:

> This can be achieved by setting the end time in the past. But since Lockup [requires end time to be in the future](https://github.com/sablier-labs/v2-core/blob/daeac7c23942ac465fd5da5d8291ecd433326e72/src/libraries/Helpers.sol#L175-L179), we will have to change that and allow end time to be in the past as long as the following inequality holds: startTime < cliffTime < endTime.

I'd love to hear your concern on allowing end time to be in the past.

**Reply by andreivladbrg** [reply]
Shees, I've missed this part, sorry

Actually, in the first version of the contracts (before first audit) we allowed to have an end time in the past, but this was one of their findings, please see [this](https://github.com/cantinasec/review-sablier/issues/7)

<details><summary>Issue</summary>
<p>

**Description:** When a new stream is created, there are no checks on the `startTime` or `endTime` parameters relative to the current time. All checks are only relative to each other.

This opens up a dangerous avenue for user error, where incorrectly entering an endTime could result in a stream that is immediately fully vested.

In such a situation, even canceling would not help, as the full value of the stream would be transferred to the recipient.

In order to help prevent this kind of user error, it is recommended to ensure that `endTime > block.timestamp`.

**Recommendation:** In the `Helpers.sol` functions that check the time params, include the following check:


</p>
</details> 

---

Now to give a full response.

As far as I understand, the motivation behind this proposal is to achieve a one-time airdrop without implementing a dedicated contract for it, and to do so using the Linear Lockup contract. Correct me if I am wrong.

The problem I see with using `createWithTimestamps` is that having the same end time for each claimer (for a non-one-time airdrop) is arguably the same as having a one-time airdrop, since the full funds will be unlocked for everyone at the same moment. Yes, you can say they can still withdraw some tokens until the end time has passed, so that you have multiple selling/buying points, but this does not offer the same dynamism as setting the start time at the claim moment, which creates unique patterns for releasing airdrop tokens (it’s very improbable that all recipients will claim at the same time).

Now, regarding the use of Lockup contracts in a one-time airdrop, yes, we could implement a `createAndWithdraw` so that the same flow (only if we change the core to allow an end time in the past) is achieved as a regular airdrop. However, this is not efficient at all; why involve a “streaming” contract just to make a simple `ERC20` transfer?

It’s like going around your elbow to get to your thumb.

<img width="300" alt="image" src="https://github.com/user-attachments/assets/87ed0de3-c1ce-4986-ba3b-5b192662248e">

My conclusion for now is that having a dedicated contract for one-time airdrops would be better. Also, I am generally against the idea of making changes to the `core` just to achieve something more easily or efficiently on the `periphery`, tho here it is not more efficient.

**Reply by smol-ninja** [reply]
> Actually, in the first version of the contracts (before first audit) we allowed to have an end time in the past, but this was one of their findings, please see https://github.com/cantinasec/review-sablier/issues/7

Thanks for sharing. After reviewing the issue, I believe it should not be acknowledged as an issue at all. Even in the current design, users can mistakenly set an end time to be just one second in the future, resulting in immediate vesting. These are user errors. Restricting end times to the future limits many use cases, as I mentioned in the original post.

---------------------

> As far as I understand, the motivation behind this proposal is to achieve a one-time airdrop without implementing a dedicated contract for it, and to do so using the Linear Lockup contract. Correct me if I am wrong.

No, I'm not against having a dedicated one-time airdrop contract, and I've started a discussion [here](https://github.com/sablier-labs/v2-periphery/discussions/370). While this proposal could be used for a one-time airdrop, a bespoke contract is preferable for gas efficiency. The goal of this discussion is to enable ranged airstreams, one-time airdrops, and legacy streams simultaneously. Currently, the design only supports legacy streams.

> The problem I see with using createWithTimestamps is that having the same end time for each claimer (for a non-one-time airdrop) is arguably the same as having a one-time airdrop, since the full funds will be unlocked for everyone at the same moment

No, they are not the same. Ranged airstreams have been used by many teams for airdrops ([Airdrops through Coinlist](https://coinlist.co/bondex) are always ranged, Avalanche token sale was also ranged facilitated by [tokensoft](https://www.tokensoft.io/)).

The reason for this discussion is that replacing `createWithDurations` with `createWithTimestamps` to unlock more Airstreams' potential without losing any current functionality.

<img width="356" alt="Screenshot 2024-07-19 at 12 01 10" src="https://github.com/user-attachments/assets/cfa604d8-314d-4f84-b495-314ab0986030">

Now that I have stated my motivation, I look forward to your criticism once again.

**Reply by andreivladbrg** [reply]
I have to admit, I’ve misunderstood the legacy airstreams point until now. Thanks for re-explaining, and I now agree with you that `createWithTimestamps` would be better than `createWithDurations`.

**Comment by andreivladbrg** [comment]
Thanks for opening this; I am in favor of this proposal.

Question: This change would be only for linear, right? For tranched, we can’t use `createWithTimestamps`, right?

**Reply by smol-ninja** [reply]
I used Lockup Linear as an example for simplicity, but the same idea applies to Lockup Tranched by allowing the last tranche timestamp to be in the past. Ranged airstreams can also have tranches.

Similarly in lockup dynamic, the last segment timestamp can be allowed to be in the past. Therefore, a general statement would be that any lockup stream can have an end time in the past.

**Reply by razgraf** [reply]
I, too, like where this is going!

**Comment by PaulRBerg** [comment]
## Initial Understanding

Thanks Shub for this. It's an interesting proposal, but I have reservations.

Am I correct that this would entail a change in the Core contracts, too? This is what you meant here, right:

> we will have to change that and allow end time to be in the past as long as the following inequality holds: startTime < cliffTime < endTime.

If yes, that's not good. Disallowing end times in the past served as a way to prevent user errors:

https://github.com/cantinasec/review-sablier/issues/7

## Further Criticism

- The name of `MerkleLockup.Timestamps` is misleading. The meme below captures my reservations better than any words could.
- Lockup is architected the way it is for a purpose - to enable onchain vesting/ distribution of assets. We shouldn't alter the core protocol to implement a periphery feature.

<img width="300" src="https://github.com/user-attachments/assets/7ce8be5f-df0b-43a3-9571-ee7af6d1d2e4"/>

## Suggestions

- Implement a separate one-time airdrop contract
- Implement your proposal to enable ranged airstreams, but find a different name for `MerkleLockup.Timestamps`

**Reply by smol-ninja** [reply]
> Am I correct that this would entail a change in the Core contracts, too? If yes, that's not good. Disallowing end times in the past served as a way to prevent user errors: https://github.com/cantinasec/review-sablier/issues/7

Yes. As I mentioned [it to Andrei](https://github.com/sablier-labs/v2-periphery/discussions/369#discussioncomment-10094625), I believe it should not be acknowledged as an issue at all. Even in the current design, users can mistakenly set an end time to be just one second in the future, resulting in immediate vesting. These are user errors. Restricting end times to the future limits many use cases, as I mentioned in the original post.

----------------

> The name of `MerkleLockup.Timestamps` is misleading

I agree and we can come up with new names. I used `Timestamps` as a placeholder. (PS: the meme is really good 🤣)

> We shouldn't alter the core protocol to implement a periphery feature

Yes and I agree with this principle. However, by not allowing `endTime` to be in the past, it restricts many use cases including [ranged airstreams](https://github.com/sablier-labs/v2-periphery/discussions/279).

I'm not against having a separate one-time airdrop contract. While this solution may be used for a one-time airdrop, a bespoke contract is preferable for gas efficiency. The goal of this discussion is to enable ranged airstreams, one-time airdrops, and legacy streams simultaneously. Currently, the design only supports legacy streams

By replacing `createWithDurations` with `createWithTimestamps`, we can unlock more Airstreams' potential without losing any current functionality.

**Reply by PaulRBerg** [reply]
> Even in the current design, users can mistakenly set an end time to be just one second in the future

That's a good point!

> I agree and we can come up with new names

Great!

> by not allowing endTime to be in the past, it restricts many use cases including https://github.com/sablier-labs/v2-periphery/discussions/279.

Another good point.

Though it's ranged *airstreams*, not *streams*. A subtle but important difference.

---

Ok so here's my updated proposal:

- Switch to `createWithTimestamps` and the Elephant-Penguin type that blends durations and timestamps into a new struct
- Update Core to allow end time in the past to enable ranged airstreams
- Build a bespoke airdrop contract for increased gas efficiency

**Reply by smol-ninja** [reply]
Cool. @andreivladbrg if you agree as well, I will create an issue in core (to allow end time to be in the past) and an issue in periphery (Switch to `createWithTimestamps`).

**Reply by andreivladbrg** [reply]
Go for it!

**Reply by PaulRBerg** [reply]
Beautiful alignment

**Comment by smol-ninja** [comment]
### Linked issues
1. https://github.com/sablier-labs/v2-core/issues/982
2. https://github.com/sablier-labs/v2-core/issues/996
